### PR TITLE
feat: L1Sload precompile

### PR DESCRIPTION
**Pull request by vyzo**
## 1. Purpose or design rationale of this PR

da precompile.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [x] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: core/vm/contracts.go [hunk 1]
```diff
@@ -125,6 +126,23 @@ var PrecompiledContractsBernoulli = map[common.Address]PrecompiledContract{
...
+		common.BytesToAddress([]byte{1}):  &ecrecover{},
+		common.BytesToAddress([]byte{2}):  &sha256hash{},
+		common.BytesToAddress([]byte{3}):  &ripemd160hashDisabled{},
+		common.BytesToAddress([]byte{4}):  &dataCopy{},
+		common.BytesToAddress([]byte{5}):  &bigModExp{eip2565: true},
+		common.BytesToAddress([]byte{6}):  &bn256AddIstanbul{},
+		common.BytesToAddress([]byte{7}):  &bn256ScalarMulIstanbul{},
+		common.BytesToAddress([]byte{8}):  &bn256PairingIstanbul{},
+		common.BytesToAddress([]byte{9}):  &blake2FDisabled{},
+		common.BytesToAddress([]byte{20}): &l1sload{l1client: cfg.L1Client},
```
Comment (Thegaram) [inline core/vm/contracts.go]: Final contract address TBD

File: params/config.go [hunk 2]
```diff
@@ -746,6 +747,11 @@ func (c *ChainConfig) IsCurie(num *big.Int) bool {
 	return isForked(c.CurieBlock, num)
 }
 
+// IsCurie returns whether num is either equal to the Curie fork block or greater.
+func (c *ChainConfig) IsDescartes(num *big.Int) bool {
```
Comment (Thegaram) [inline params/config.go]: These were added 3 weeks ago in https://github.com/scroll-tech/go-ethereum/pull/698. Please don't re-add them here, but merge `develop` to your target branch.

File: core/vm/contracts.go [hunk 3]
```diff
@@ -1130,3 +1155,17 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
...
+	l1client L1Client
+}
+
+func (c *l1sload) RequiredGas(input []byte) uint64 {
+	// TODO
+	return 1234567
+}
+
+func (c *l1sload) Run(input []byte) ([]byte, error) {
+	return c.l1client.SLoad(context.TODO(), input)
```
Comment (Thegaram) [inline core/vm/contracts.go]: Better to parse `input` here and pass the actual arguments to `l1client`.

File: core/vm/contracts.go (hunk 1 above)
Comment (vyzo) [reply core/vm/contracts.go]: right, will add TODO

File: params/config.go (hunk 2 above)
Comment (vyzo) [reply params/config.go]: uh ok, better to merge to the base branch and rebase thr pr.

File: core/vm/contracts.go (hunk 3 above)
Comment (vyzo) [reply core/vm/contracts.go]: agreed, esp. given that we have to validate thr height.

File: core/vm/contracts.go [hunk 4]
```diff
@@ -130,6 +131,10 @@ var PrecompiledContractsBernoulli = map[common.Address]PrecompiledContract{
 // PrecompiledContractsDescartes returns the default set of precompiled contracts,
 // including the L1SLoad precompile.
 func PrecompiledContractsDescartes(cfg Config) map[common.Address]PrecompiledContract {
+	if cfg.L1Client == nil || (reflect.ValueOf(cfg.L1Client).Kind() == reflect.Ptr && reflect.ValueOf(cfg.L1Client).IsNil()) {
+		// TODO: have a better way to handle this error
+		panic("No L1 client provided in the vm config")
+	}
```
Comment (vyzo) [inline core/vm/contracts.go]: just a nil check will do, the reflect business is really unnecessary.

Comment (icemelon) [reply core/vm/contracts.go]: fixed

File: params/config.go (hunk 2 above)
Comment (icemelon) [reply params/config.go]: merged the latest develop branch

**Pull request merged**